File: Algorithms/Copy List with Random Pointer.py
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if head==None:
            return head
        copy=Node(head.val)
        temp=head.next
        c=copy
        while(temp):
            c.next=Node(temp.val)
            c=c.next
            temp=temp.next
        temp=head
        c=copy
        while(temp):
            next=temp.next
            temp.next=c
            c.random=temp.random
            c=c.next
            temp=next
        c=copy
        while(c):
            if c.random!=None:
                c.random=c.random.next
            c=c.next
        return copy

File: Algorithms(Java)/Copy List with Random Pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if (head==null){
            return head;
        }
        Node copy=new Node(-1);
        Node temp=head;
        Node c=copy;
        HashMap<Node,Node> h=new HashMap<Node,Node>();
        while(temp!=null){
            c.next=new Node(temp.val);
            h.put(temp,c.next);
            c=c.next;
            temp=temp.next;
        }
        temp=head;
        c=copy.next;
        while(temp!=null){
            c.random=temp.random!=null?h.get(temp.random):null;
            temp=temp.next;
            c=c.next;
        }
        return copy.next;
    }
}